Framework/Algorithms/src/MonteCarloAbsorption.cpp

```diff
@@ -416,30 +417,39 @@ MatrixWorkspace_uptr MonteCarloAbsorption::createOutputWorkspace(const MatrixWor
  * @return A new IBeamProfile object
  */
 std::unique_ptr<IBeamProfile> MonteCarloAbsorption::createBeamProfile(const Instrument &instrument,
-                                                                      const IObject &sample) const {
+                                                                      const Sample &sample) const {
   const auto frame = instrument.getReferenceFrame();
   const auto source = instrument.getSource();
 
-  std::string beamShapeParam = source->getParameterAsString("beam-shape");
+  const std::string beamShapeParam = source->getParameterAsString("beam-shape");
   if (beamShapeParam == "Slit") {
-    auto beamWidthParam = source->getNumberParameter("beam-width");
-    auto beamHeightParam = source->getNumberParameter("beam-height");
+    const auto beamWidthParam = source->getNumberParameter("beam-width");
+    const auto beamHeightParam = source->getNumberParameter("beam-height");
     if (beamWidthParam.size() == 1 && beamHeightParam.size() == 1) {
       return std::make_unique<RectangularBeamProfile>(*frame, source->getPos(), beamWidthParam[0], beamHeightParam[0]);
     }
   } else if (beamShapeParam == "Circle") {
-    auto beamRadiusParam = source->getNumberParameter("beam-radius");
+    const auto beamRadiusParam = source->getNumberParameter("beam-radius");
     if (beamRadiusParam.size() == 1) {
       return std::make_unique<CircularBeamProfile>(*frame, source->getPos(), beamRadiusParam[0]);
     }
   } // revert to sample dimensions if no return by this point
-  if (!sample.hasValidShape())
-    throw std::invalid_argument("Cannot determine beam profile without a sample shape");
-  const auto bbox = sample.getBoundingBox().width();
-  const auto bboxCentre = sample.getBoundingBox().centrePoint();
-  const double beamWidth = 2 * bboxCentre[frame->pointingHorizontal()] + bbox[frame->pointingHorizontal()];
-  const double beamHeight = 2 * bboxCentre[frame->pointingUp()] + bbox[frame->pointingUp()];
-  return std::make_unique<RectangularBeamProfile>(*frame, source->getPos(), beamWidth, beamHeight);
+  if (!sample.getShape().hasValidShape() && !sample.hasEnvironment()) {
+    throw std::invalid_argument("Cannot determine beam profile without a sample shape and environment");
+  } else {
+    V3D bbox;
```

Comment:
We should omit this else branch and bring everything up one level since the throw above makes the else obsolete 
This feels like a nit, but clang-tidy also warns for this too so it's saving future work

Reply:
So this:
```c++
  if (!sample.getShape().hasValidShape() && !sample.hasEnvironment()) {
    throw std::invalid_argument("Cannot determine beam profile without a sample shape and environment");
  } else {
    V3D bbox;
    V3D bboxCentre;
    if (sample.getShape().hasValidShape()) {
      bbox = sample.getShape().getBoundingBox().width();
      bboxCentre = sample.getShape().getBoundingBox().centrePoint();
    } else {
      bbox = sample.getEnvironment().boundingBox().width();
      bboxCentre = sample.getEnvironment().boundingBox().centrePoint();
    }
    const double beamWidth = 2 * bboxCentre[frame->pointingHorizontal()] + bbox[frame->pointingHorizontal()];
    const double beamHeight = 2 * bboxCentre[frame->pointingUp()] + bbox[frame->pointingUp()];
    return std::make_unique<RectangularBeamProfile>(*frame, source->getPos(), beamWidth, beamHeight);
  }
}
```

Can be rewritten as:
```c++
  if (!sample.getShape().hasValidShape() && !sample.hasEnvironment()) {
    throw std::invalid_argument("Cannot determine beam profile without a sample shape and environment");
  }
  V3D bbox;
  V3D bboxCentre;
  if (sample.getShape().hasValidShape()) {
    bbox = sample.getShape().getBoundingBox().width();
    bboxCentre = sample.getShape().getBoundingBox().centrePoint();
  } else {
    bbox = sample.getEnvironment().boundingBox().width();
    bboxCentre = sample.getEnvironment().boundingBox().centrePoint();
  }
  const double beamWidth = 2 * bboxCentre[frame->pointingHorizontal()] + bbox[frame->pointingHorizontal()];
  const double beamHeight = 2 * bboxCentre[frame->pointingUp()] + bbox[frame->pointingUp()];
  return std::make_unique<RectangularBeamProfile>(*frame, source->getPos(), beamWidth, beamHeight);
}
```

Since the throw causes the stack to unwind at the point of throw, so we can't get into the `else` block below making it redundant.
Clang-tidy highlights this as it usually causes us to have 2/3 nested if-else blocks (like this case) and can be bugprone so we remove the else to make the control flow easier to read-through

Reply:
Now I get it, I thought that instead of flattening you are suggesting removing the case altogether. The last commit implements your suggestion.